### PR TITLE
Fix issue #5 on branch core-api-v1

### PR DESCRIPTION
**Pull request by fbaligand**
Fix issue #5.
When code call raises an exception, the error is logged and the event is tagged '_aggregateexception'.
Thus, it avoids logstash crash.

### CODE REVIEW

File: lib/logstash/filters/aggregate.rb [hunk 1]
```diff
@@ -202,13 +215,20 @@ def filter(event)
 			map = aggregate_maps_element.map
 
 			# execute the code to read/update map and event
-			@codeblock.call(event, map)
+			begin
+				@codeblock.call(event, map)
+				noError = true
+			rescue Exception => exception
```
Comment (purbon) [inline lib/logstash/filters/aggregate.rb]: Hi,
  is actually more idiomatic to rescue StandardError descendant exceptions here. You can use something like:

```
  begin 
     ....
  rescue => e
  end
```

If you want more details I found this article on the internet: https://robots.thoughtbot.com/rescue-standarderror-not-exception

**Comment by purbon** [comment]
@fbaligand nice job here man! I would also recommend you add a few test to handle this issue. You could for example mock the block call and make it raise an exception. Like this we would catch this in te feature. Does this makes sense to you?

File: lib/logstash/filters/aggregate.rb (hunk 1 above)
Comment (fbaligand) [inline lib/logstash/filters/aggregate.rb]: OK, no problem for that. I will fix it soon.

Just for information, `rescue Exception => exception` was inspired from this code : 
https://github.com/logstash-plugins/logstash-filter-ruby/blob/master/lib/logstash/filters/ruby.rb#L40

=> Maybe it needs also a fix ?

**Comment by fbaligand** [comment]
I agree it would be better to make a test ;)
I will do it.

File: lib/logstash/filters/aggregate.rb (hunk 1 above)
Comment (purbon) [inline lib/logstash/filters/aggregate.rb]: sure, also not much good :+1: thanks man.

**Comment by purbon** [comment]
thanks a lot @fbaligand good job!

**Comment by fbaligand** [comment]
- Exception handling has been fixed
- A unit test has been added to check aggregate plugin behaviour when an error occurs

=> @purbon Is it OK for you ?

**Comment by fbaligand** [comment]
Hi @purbon,

Have you reviewed my fixes ? Is it OK for you ?

**Comment by fbaligand** [comment]
Hi @purbon,

Can you review my fixes ?
If you need something more, tell me.

**Comment by purbon** [comment]
this will need to be ported to master, for next times I think we should always focus on doing changes to master and then backporting, for this one I agree is easy to just move to the feature :P  ... let's review this and then port it to master (2.0 branch).

File: lib/logstash/filters/aggregate.rb [hunk 2]
```diff
@@ -202,13 +216,20 @@ def filter(event)
...
+				event.tag("_aggregateexception")
+			end
 			
 			# delete the map if task is ended
 			@@aggregate_maps.delete(task_id) if @end_of_task
 		end
 
-		filter_matched(event)
+		# match the filter, only if no error occurred
+		filter_matched(event) if noError
```
Comment (purbon) [inline lib/logstash/filters/aggregate.rb]: do you need the variable here? or could you probably just check if the event has the tag?

**Comment by purbon** [comment]
@fbaligand just leave you an small comment, except from that, test pass and code looks good to me. If we agree on the comment, I think we are cool to move forward.

File: lib/logstash/filters/aggregate.rb (hunk 2 above)
Comment (fbaligand) [inline lib/logstash/filters/aggregate.rb]: Check if the event has the _aggregateerror tag may be a bad idea in some case.
I mean : if the event passed several aggregate filters, and a previous aggregate filter raised an error, but not current one, then the event has the _aggregateerror tag but we still want to call filter_matched() method.

What do you think about it ?

Comment (purbon) [inline lib/logstash/filters/aggregate.rb]: make sense, what do you think about synchronization issues with this variable?

Comment (fbaligand) [inline lib/logstash/filters/aggregate.rb]: @purbon 
"noError" is a local variable in method "filter".
So there is no synchronization issue with this variable.

Comment (purbon) [inline lib/logstash/filters/aggregate.rb]: makes sense, sorry busy day doing too many things at once.

**Comment by purbon** [comment]
LGTM, I guess we can squash and start the porting to master, ok for you?

**Comment by fbaligand** [comment]
This is ok for me. 
Who is porting to master ? You or me ?

**Comment by purbon** [comment]
If you squash the commit, I'm happy to do it myself, or you can do it yourself. Your call.

**Comment by fbaligand** [comment]
I merged the branch core-api-v1 to master.

2015-10-12 21:49 GMT+02:00 Pere Urbón <notifications@github.com>:

**Comment by fbaligand** [comment]
@purbon 
You gave the LGTM... but it's still not merged :)
Do you wait for something ?

**Comment by purbon** [comment]
no, will merge soon.

- purbon

**Comment by fbaligand** [comment]
OK, thanks for the reply.

**Comment by fbaligand** [comment]
Thank you :)

2015-10-14 13:44 GMT+02:00 Pere Urbón <notifications@github.com>:

**Comment by purbon** [comment]
@fbaligand sorry took some time, this is something we're working on to fix soon.

**Comment by fbaligand** [comment]
@purbon 
No problem :)

**Pull request merged**